File: bukkit/libs/misc/src/main/java/dialight/observable/map/ObservableMap.java
package dialight.observable.map;

import dialight.function.A3Consumer;
import dialight.observable.collection.ObservableCollection;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Function;

public abstract class ObservableMap<K, V> implements Map<K, V> {

    private final Map<Object, BiConsumer<K, V>> onPut = new LinkedHashMap<>();
    private final Map<Object, BiConsumer<K, V>> onRemove = new LinkedHashMap<>();
    private final Map<Object, A3Consumer<K, V, V>> onReplace = new LinkedHashMap<>();


    protected void firePut(K k, V v) {
        for(BiConsumer<K, V> op : onPut.values()) op.accept(k, v);
    }
    protected void fireRemove(K k, V v) {
        for(BiConsumer<K, V> op : onRemove.values()) op.accept(k, v);
    }
    protected void fireReplace(K k, V oldValue, V newValue) {
        for(A3Consumer<K, V, V> op : onReplace.values()) op.apply(k, oldValue, newValue);
    }

    public ObservableMap<K, V> onPut(Object key, BiConsumer<K, V> op) {
        onPut.put(key, op);
        return this;
    }
    public ObservableMap<K, V> onRemove(Object key, BiConsumer<K, V> op) {
        onRemove.put(key, op);
        return this;
    }
    public ObservableMap<K, V> onReplace(Object key, A3Consumer<K, V, V> op) {
        onReplace.put(key, op);
        return this;
    }

    public ObservableMap<K, V> removeListeners(Object key) {
        onPut.remove(key);
        onRemove.remove(key);
        onReplace.remove(key);
        return this;
    }

//    public abstract V put(K key, V value);
//    public abstract V remove(K key);

    public ObservableCollection<V> asCollectionObaervable(Function<V, K> getKey) {
        return new ValuesObservable<>(this, getKey);
    }
    public ObservableCollection<V> asImmutableCollectionObaervable(Function<V, K> getKey) {
        return new ValuesImmutableObservable<>(this, getKey);
    }

}

File: sponge/libs/misc/src/main/kotlin/dialight/observable/map/ObservableMap.kt
package dialight.observable.map

import java.util.*

abstract class ObservableMap<K, V> {

    private val onPut: MutableCollection<(K, V) -> Unit> = LinkedList()
    private val onRemove: MutableCollection<(K, V) -> Unit> = LinkedList()
    private val onReplace: MutableCollection<(K, V, V) -> Unit> = LinkedList()


    protected fun firePut(k: K, v: V) {
        for(op in onPut) op(k, v)
    }
    protected fun fireRemove(k: K, v: V) {
        for(op in onRemove) op(k, v)
    }
    protected fun fireReplace(k: K, old: V, new: V) {
        for(op in onReplace) op(k, old, new)
    }


    fun onPut(op: (K, V) -> Unit) : ObservableMap<K, V> {
        onPut += op
        return this
    }
    fun onRemove(op: (K, V) -> Unit) : ObservableMap<K, V> {
        onRemove += op
        return this
    }
    fun onReplace(op: (key: K, old: V, new: V) -> Unit) : ObservableMap<K, V> {
        onReplace += op
        return this
    }

    abstract fun put(key: K, value: V): V?
    abstract fun remove(key: K): V?

}